[JSC] Adopt Intl.NumberFormat v3 spec change for useGrouping option

Constellation opened the pull request:
#### 70694b2a29d9e4233af0341e709582086740785d
<pre>
[JSC] Adopt Intl.NumberFormat v3 spec change for useGrouping option
<a href="https://bugs.webkit.org/show_bug.cgi?id=242200">https://bugs.webkit.org/show_bug.cgi?id=242200</a>
rdar://93712843

Reviewed by Ross Kirsling.

Intl.NumberFormat v3&apos;s useGrouping option has incompatibility with the previous version.
We observe it in JSTests and PerformanceTests. And we discussed it in ECMA-402 meeting[1].
ECMA-402 members agree that we fix this incompatibility, and it is done in [2].
This patch applies this change to our implementation.

[1]: <a href="https://github.com/tc39/ecma402/blob/master/meetings/notes-2022-03-17.md#web-compatibility-allow-the-string-true-for-usegrouping">https://github.com/tc39/ecma402/blob/master/meetings/notes-2022-03-17.md#web-compatibility-allow-the-string-true-for-usegrouping</a>
[2]: <a href="https://github.com/tc39/proposal-intl-numberformat-v3/commit/0cf396523575aab0a9313ebc59cdb09af25781f1">https://github.com/tc39/proposal-intl-numberformat-v3/commit/0cf396523575aab0a9313ebc59cdb09af25781f1</a>

* JSTests/stress/intl-numberformat-usegrouping-v3.js:
(nonListedUseGrouping.forEach):
(invalidUseGrouping.forEach): Deleted.
* JSTests/test262/expectations.yaml:
* PerformanceTests/Skipped:
* Source/JavaScriptCore/runtime/IntlNumberFormat.cpp:
(JSC::IntlNumberFormat::initializeNumberFormat):
* Source/JavaScriptCore/runtime/IntlObjectInlines.h:
(JSC::intlStringOrBooleanOption):

Canonical link: <a href="https://commits.webkit.org/252013@main">https://commits.webkit.org/252013@main</a>
</pre>

rkirsling commented on Source/JavaScriptCore/runtime/IntlObjectInlines.h:
```diff
@@ -167,25 +167,24 @@ ResultType intlStringOrBooleanOption(JSGlobalObject* globalObject, JSObject* opt
 
-        bool valueBoolean = value.toBoolean(globalObject);
-        RETURN_IF_EXCEPTION(scope, { });
+    if (value.isBoolean() && value.asBoolean())
+        return trueValue;
```
Shouldn't we handle explicit `false` here too?
```suggestion
    if (value.isBoolean())
        return value.asBoolean() ? trueValue : falsyValue;
```

Constellation replied:
Oops! Correct. Fix, and add tests.

Constellation replied:
Ah, no. According to the spec change, we are delegating this evaluation to `if (!valueBoolean)`.

```
1. If _value_ is *undefined*, return _fallback_.
1. If _value_ is *true*, return _trueValue_.
1. Let _valueBoolean_ be ToBoolean(_value_).
1. If _valueBoolean_ is *false*, return _falsyValue_.
1. Let _value_ be ? ToString(_value_).
1. If _values_ does not contain an element equal to _value_, return _fallback_.
1. Return _value_.
```

rkirsling replied:
Either way should be correct, I just thought it might be a little nicer as code. 😅

Constellation replied:
Ah, I see. I think probably the current code is fine since it is well aligned to the spec text ;)

rkirsling reviewed (approved):
r=me with a comment

Constellation committed:
[JSC] Adopt Intl.NumberFormat v3 spec change for useGrouping option
https://bugs.webkit.org/show_bug.cgi?id=242200
rdar://93712843

Reviewed by Ross Kirsling.

Intl.NumberFormat v3's useGrouping option has incompatibility with the previous version.
We observe it in JSTests and PerformanceTests. And we discussed it in ECMA-402 meeting[1].
ECMA-402 members agree that we fix this incompatibility, and it is done in [2].
This patch applies this change to our implementation.

[1]: https://github.com/tc39/ecma402/blob/master/meetings/notes-2022-03-17.md#web-compatibility-allow-the-string-true-for-usegrouping
[2]: https://github.com/tc39/proposal-intl-numberformat-v3/commit/0cf396523575aab0a9313ebc59cdb09af25781f1

* JSTests/stress/intl-numberformat-usegrouping-v3.js:
(nonListedUseGrouping.forEach):
(invalidUseGrouping.forEach): Deleted.
* JSTests/test262/expectations.yaml:
* PerformanceTests/Skipped:
* Source/JavaScriptCore/runtime/IntlNumberFormat.cpp:
(JSC::IntlNumberFormat::initializeNumberFormat):
* Source/JavaScriptCore/runtime/IntlObjectInlines.h:
(JSC::intlStringOrBooleanOption):

Canonical link: https://commits.webkit.org/252013@main
